actionpack/lib/action_dispatch/middleware/stack.rb

```diff
@@ -102,11 +105,48 @@ def build(app = Proc.new)
     private
 
       def assert_index(index, where)
-        i = index.is_a?(Integer) ? index : middlewares.index { |m| m.klass == index }
+        if index.is_a?(Integer)
+          i = index
+        else
+          if target_deleted?(index)
+            target = next_target(index, where)
+
+            return -1 unless target
+          else
+            target = index
+          end
+
+          i = middleware_index(target)
+        end
+
         raise "No such middleware to insert #{where}: #{index.inspect}" unless i
+
         i
       end
 
+      def target_deleted?(target)
+        @deleted_middlewares.key?(target)
+      end
+
+      def next_target(target, where)
+        position = where == :after ? :previous : :next
+
+        @deleted_middlewares[target][position]
+      end
+
+      def middleware_index(klass)
+        middlewares.index { |m| m.klass == klass }
+      end
+
+      def track_deleted_middleware_at(index)
+        previous_middleware = middlewares[index - 1].klass
```

Comment:
This doesn't work well if you delete the very first middleware.  In such a case, the `previous` pointer will point to the _last_ middleware in the stack, so subsequent calls to insert relative to that pointer will put things near the end rather than near the beginning, for example.